modules/qna/src/backend/storage.ts

```diff
@@ -119,23 +121,38 @@ export default class Storage {
     }
   }
 
-  async insert(qna: QnaEntry | QnaEntry[]): Promise<string[]> {
-    const ids = await Promise.mapSeries(_.isArray(qna) ? qna : [qna], async (data, i) => {
-      const id = makeID(data)
-      await this.checkForDuplicatedQuestions(data, id)
-      const item: QnaItem = { id, data }
-      if (data.enabled) {
-        await this.createNLUIntentFromQnaItem(item, true)
-      }
-
-      await this.bp.ghost
-        .forBot(this.botId)
-        .upsertFile(this.config.qnaDir, `${id}.json`, JSON.stringify(item, undefined, 2))
+  async upsertItem(item: QnaItem | QnaItem[]): Promise<string[]> {
+    const items = _.isArray(item) ? item : [item]
+    const qnaMap: { [key: string]: QnaEntry } = {}
 
-      return id
+    items.forEach(async item => {
+      if (item.id in qnaMap) {
+        this.bp.logger.warn(`Duplicate IDs found in input while batch importing: ${item.id}`)
+        qnaMap[makeID(item.data)] = item.data
+      }
+      qnaMap[item.id] = item.data
     })
 
-    return ids
+    return Promise.all(
+      Object.entries(qnaMap).map(async ([id, data]) => {
+        await this.checkForDuplicatedQuestions(data, id)
+        const item: QnaItem = { id, data }
+        if (data.enabled) {
+          await this.createNLUIntentFromQnaItem(item, true)
+        }
+
+        await this.bp.ghost
+          .forBot(this.botId)
+          .upsertFile(this.config.qnaDir, `${id}.json`, JSON.stringify(item, undefined, 2))
+        return id
+      })
+    )
+  }
+
+  async insert(data: QnaEntry): Promise<string> {
+    const id = makeID(data)
+    const createdIds = await this.upsertItem({ id, data })
+    return createdIds[0]
```

Comment:
createdId[0] will be the same as id right ?

Reply:
yes